runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PCollectionTranslation.java

```diff
@@ -33,7 +34,10 @@ private PCollectionTranslation() {}
 
   public static RunnerApi.PCollection toProto(PCollection<?> pCollection, SdkComponents components)
       throws IOException {
-    String coderId = components.registerCoder(pCollection.getCoder());
+    String coderId = components.registerCoder(
```

Comment:
This is wrong. The PCollection's coder shouldn't be wrapped like this.

Reply:
Right now there's an inconsistency the meaning of the coder field of a PCollection depending on whether the proto is being used in a runner or a execution context, and this seems even more wrong. At execution time the runner does need to coordinate with the SDK on how to represent windowed elements, and there are multiple valid representations. Thus I think we must have it everywhere rather than nowhere.

Reply:
* The particular encoding used on the data plane is part of the particular invocation over the Fn API, not the pipeline.
 * The WindowedValue with an element and a list of windows is a compressed representation of multiple elements in their respective windows.
 * The SDK harness should not leak a representation to user logic.

If I understand the situation today, the FullWindowedValueCoder is always used. We should allow this to evolve, and allow runners to optimize based on ParDoPayloads, encoding and transmitting only bits that are needed if the runner determines it pays off versus reencoding costs, etc.

Reply:
IMO none of this is new - the use of WindowedValueCoder in the DataflowRunner and elsewhere is just an artifact of imposing the Beam model on runners that don't natively understand it.

Reply:
I'm not sure what you're proposing here. We shouldn't be re-using a field to have a different meaning in different contexts.

Reply:
The Fn API reusing the Pipeline protos is just a way to have a common graph representation. The pipeline's coder indicates the coder of elements in the PCollection (the obvious thing?) and at least right now that field is used to indicate the encoding of an element with windows and metadata on the wire of the data plane.

Reply:
FWIW it is you who are proposing something :-)

Reply:
Fair point, I'm proposing eliminating inconsistency in how these fields are used. I didn't imagine this inconsistency was a conscious choice. 

I suppose we could add yet another field, windowed_element_coder (or call this something other than "coder" if we want to avoid that) that's generally only populated by the runner and used during fn execution. Perhaps this is where the len_prefix variations would go as well. Perhaps it should be tied to the data plane/state API itself rather than the PCollections? 

Regarding sending a list of Windows, this is an example of a special representation that's an important optimization for sliding windows shuffle data reduction. (Yes, we just happen to use it everywhere...)

Reply:
My personal preference would be to tie it to the data plane, or I guess control plane since the data plane is just the byte stream. In some ways, PCollection is a misnomer for the intermediate nodes of a ProcessBundleDescriptor. On the other hand, having a single graph structure is nice.

And I am not at all against this special representation on the wire - I agree it is a huge win. I just want to leave the door open to more wins. Like single-firing bounded global window computation, where we can omit all the metadata. I think each of them will need to be standardized so SDKs can implement them, and the runner chooses from a menu of options. Going into fancier negotiation protocols seems excessive for this particular case since there's probably a single-digit number of `Coder<WindowedValue>` that are interesting. OTOH a pretty interesting case is going columnar so there's no `WindowedValue` to speak of at all.